src/microbe_stage/gui/HUDMessages.cs

```diff
@@ -0,0 +1,162 @@
+    public Color BaseMessageColour = new(1, 1, 1);
+
+    [Export]
+    public int MaxShownMessages = 4;
+
+    [Export]
+    public float MaxDisplayTimeBeforeForceFade = 60;
+
+    /// <summary>
+    ///   The alpha value that is reached by the midway point of a message
+    /// </summary>
+    [Export]
+    public float MidwayFadeValue = 0.65f;
+
+    private readonly List<(IHUDMessage Message, Label Displayer)> hudMessages = new();
+
+    private string multipliedMessageTemplate = string.Empty;
+
+    public override void _Ready()
+    {
+        multipliedMessageTemplate = TranslationServer.Translate("HUD_MESSAGE_MULTIPLE");
+
+        if (MaxShownMessages < 1)
+        {
+            GD.PrintErr($"{nameof(MaxShownMessages)} needs to be at least one");
+            MaxShownMessages = 1;
+        }
+    }
+
+    public override void _Process(float delta)
+    {
+        bool clean = false;
+
+        foreach (var (message, displayer) in hudMessages)
+        {
+            message.TimeRemaining -= delta;
+
+            if (message.TimeRemaining < 0)
+            {
+                displayer.QueueFree();
+                clean = true;
+                continue;
+            }
+
+            message.TotalDisplayedTime += delta;
+
+            // Update fade
+            // TODO: should different types of messages (more urgent?) have different colours
+            displayer.SelfModulate = new Color(BaseMessageColour,
+                CalculateMessageAlpha(message.TimeRemaining, message.OriginalTimeRemaining));
+        }
+
+        if (clean)
+            hudMessages.RemoveAll(t => t.Message.TimeRemaining < 0);
+    }
+
+    public void ShowMessage(IHUDMessage message)
+    {
+        // First combine to an existing message if possible
+        foreach (var (existingMessage, existingLabel) in hudMessages)
+        {
+            if (existingMessage.IsSameMessage(message))
+            {
+                // Limit to only combining with not super long lived messages. This ensures that all messages will
+                // eventually disappear
+                if (existingMessage.TotalDisplayedTime > MaxDisplayTimeBeforeForceFade)
+                    continue;
+
+                existingMessage.UpdateFromOtherMessage(message);
+                existingLabel.Text = TextForMessage(existingMessage);
+
+                // Reset the fade time for the message that appeared again
+                existingMessage.TimeRemaining = existingMessage.OriginalTimeRemaining;
+
+                return;
+            }
+        }
+
+        // Can't combine, need to add a new label to display this
+        var label = new Label
+        {
+            SizeFlagsHorizontal = (int)SizeFlags.ExpandFill,
+            Align = Label.AlignEnum.Center,
+            Autowrap = true,
+            MouseFilter = MouseFilterEnum.Ignore,
+            Text = TextForMessage(message),
+        };
+
+        label.AddFontOverride("font", MessageFont);
+
+        AddChild(label);
```

Comment:
Could also have an option for messages to be stacked top-bottom here, where new messages are placed topmost first. I'd like to reuse this class for the kill feed messages in multiplayer.

Here's for example how I did it: https://github.com/Revolutionary-Games/Thrive/blob/e9d8965cdaa35f7f58f6e54ac4458ff8461ed353/src/microbe_stage/multiplayer/microbial_arena/MicrobialArenaHUD.cs#L166-L167

Reply:
That'll probably need other changes as well, but I'll add a reverse order option to the HUDMessages class.